Add `init()` to the `Chip` trait

bradjc opened the pull request:
### Pull Request Overview

Tock has a very ad-hoc initialization convention. Every arch/chip/board has roughly created its own procedure, with any commonalities a byproduct of convention and copying rather than anything expressed in types. This tries to add some structure by adding an `init()` function to the `Chip` trait. Boards can then all call `ChipHw::init()` at the beginning of their main() to run all the necessary chip-level initialization.

This came up in the deferred call initialization discussions. ~~Since initializing single thread value requires knowing the ThreadIdProvider, moving the deferred call init to chips is both doable and avoids a potentially very annoying mistake in boards.~~ However, we don't need to couple this with deferred call, so let's punt on that.


### Testing Strategy

todo


### TODO or Help Wanted

Thoughts?

Obviously this would need to be added to all chips.


### Documentation Updated

- [ ] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [ ] Ran `make prepush`.

bradjc committed:
kernel: add `init()` to `Chip`

This provides a standardized way for a board to run any chip-specific
setup code at kernel boot.

bradjc committed:
boards: nrf52840dk: use ChipHw::init()

bradjc committed:
chips: impl init() for Chip

For the nrf52:
Move the errata setup, vector table, and interrupt enable to functions,
document safety, and call everything from the new init() trait function.

brghena reviewed (commented):
Looking at this code, I feel like a chip init function should have always existed. That makes it feel like a pretty good change to me.

However, it feels a little strange to stuff deferred call initialization in there. Maybe there should be a kernel init function in addition to the chip init function?

bradjc commented:
> However, it feels a little strange to stuff deferred call initialization in there. Maybe there should be a kernel init function in addition to the chip init function?

Yeah, on one hand DeferredCall is a core-kernel-provided resource, on the other it is a chip-specific initialization. If we added a kernel init to kernel.rs it would be virtually a direct copy of what is in deferred_call, complete with the safe and unsafe versions and the ThreadIdProvider parameter.

We can't add deferred call init to kernel::new or kernel::loop because the init must happen essentially first. By lumping it in with a chip init 1) users don't have to remember to do it, and 2) it will come first in main.

bradjc commented:
I think this is still a promising idea. Is there an appetite for this API?

jrvanwhy commented on kernel/src/platform/chip.rs:
```diff
@@ -28,6 +28,17 @@ pub trait Chip {
+    /// - Any architecture-specific setup
+    /// - Any errata fixes
+    /// - Any configuration needed for the chip to continue booting the kernel
+    fn init();
```
Maybe this should be the constructor for the Chip (i.e. `fn init() -> Self`)?

On a related note, does this need something to prevent you from calling it multiple times?

hudson-ayers replied:
This feels overly defensive to me. I think it is easy to go too far with typestate programming and end up with really verbose code, and this check doesn't feel super valuable to me.

hudson-ayers replied:
That is about Brad's idea, I think constructor would be fine

alevy commented:
Can/should we just have a default implementation for `init`? Thus avoiding both backwards compatibility issues and the large footprint of this PR?

bradjc committed:
chips: move more init() to Chip::init()

bradjc committed:
boards: x86: call chip init

bradjc committed:
chip: lpc55s6x: use init fn

bradjc commented:
> Can/should we just have a default implementation for `init`? Thus avoiding both backwards compatibility issues and the large footprint of this PR?

I think that would reduce the changed files from 84 to about 79. Basically many riscv boards don't have an init step. Not sure it's worth it.

bradjc committed:
chips: apollo3: init: better cast

jrvanwhy commented on chips/x86_q35/src/chip.rs:
```diff
@@ -124,6 +124,14 @@ impl<I2: InterruptService, const PR: u16> Pc<'static, PcDefaultPeripherals<PR>,
+    fn init() {
+        // CPU/interrupt controller
+        unsafe {
+            x86::init();
```
`x86::init()` has the following safety comment:

```rust
/// ## Safety
///
/// This function must never be executed more than once.
///
/// [rest of safety comment skipped]
```

I'm not convinced this is sound.

bradjc replied:
This is quagmire.

I guess that pushing through the requirement to only call things once is one approach (and what is there now). But I'm not convinced that it is _unsafe_ to call things twice, rather its just not likely to work to call things twice. But, in any case there is a lot of unsafe code.

This root issue is there are a lot of static buffers shared with hardware in the x86 arch crate, that I don't think are easy to untangle.

Probably the only reasonable way forward for now is to not use this new init API for x86, and keep with the same unsafe peripherals constructor we have now.

jrvanwhy replied:
From a soundness perspective, I think that both options (pushing through the safety requirement to only call things once and making the API safe) are both okay, I'm not sure which is better.

This PR is already decently large, so leaving x86 as-is is fine with me.

jrvanwhy replied:
To clarify: When I said "leaving x86 as-is", I meant not using `Chip::init` for x86. The current iteration of this PR still looks unsound, so I don't want to merge the PR in its current state.

bradjc replied:
Reverted.

jrvanwhy commented on chips/x86_q35/src/chip.rs:
```diff
@@ -124,6 +124,14 @@ impl<I2: InterruptService, const PR: u16> Pc<'static, PcDefaultPeripherals<PR>,
+        // CPU/interrupt controller
+        unsafe {
+            x86::init();
+            crate::pic::init();
```
We need a safety comment for the `pic::init` call.

jrvanwhy commented on chips/rp2350/src/chip.rs:
```diff
@@ -50,6 +50,22 @@ impl<I: InterruptService> Chip for Rp2350<'_, I> {
+                    processor as u8
+                ),
+            }
+        }
```
Only the two `cortexm33::nvic` functions are `unsafe`, so this `unsafe` block should be smaller.

jrvanwhy commented on chips/rp2040/src/chip.rs:
```diff
@@ -60,6 +60,22 @@ impl<I: InterruptService> Chip for Rp2040<'_, I> {
+    fn init() {
+        unsafe {
+            cortexm0p::nvic::disable_all();
+            cortexm0p::nvic::clear_all_pending();
```
This `unsafe` block should only contain the two `nvic` operations.

jrvanwhy commented on chips/apollo3/src/chip.rs:
```diff
@@ -97,6 +97,31 @@ impl<I: InterruptService + 'static> Chip for Apollo3<I> {
+    fn init() {
+        unsafe {
+            let cache_ctrl = crate::cachectrl::CacheCtrl::new();
+            cache_ctrl.enable_cache();
```
The two `cache_control` lines don't need to be in the `unsafe` block.

bradjc committed:
kernel: chip: add more init comments

bradjc committed:
chips: apollo3: better init unsafe

bradjc committed:
chips: rpi: init: nvic operations are unsafe

alevy commented:
@bradjc now there's a merge conflict, but if we resolve those, I think I approve and we merge (before other conflicts arise).

alevy committed:
Merge branch 'master' into dev/chip-init

bradjc committed:
chip init: revert x86

Use the existing init practice to ensure we are not violating safety
requirements.

alevy commented:
@bradjc this ready?

bradjc commented:
Yes now that we reverted the x86 chip.
